kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/SnapshotManager.java

```diff
@@ -625,14 +599,56 @@ protected Optional<LogSegment> getLogSegmentForVersion(
                 })
             .orElse(Collections.emptyList());
 
-    return Optional.of(
-        new LogSegment(
-            logPath,
-            newVersion,
-            deltasAfterCheckpoint,
-            newCheckpointFiles,
-            newCheckpointOpt.map(x -> x.version),
-            lastCommitTimestamp));
+    return new LogSegment(
+        logPath,
+        newVersion,
+        deltasAfterCheckpoint,
+        newCheckpointFiles,
+        newCheckpointOpt.map(x -> x.version),
+        lastCommitTimestamp);
+  }
+
+  /////////////////////////
+  // getLogSegment utils //
+  /////////////////////////
+
+  /**
+   * Determine the starting checkpoint version that is at or before `versionToLoadOpt`. If no
+   * `versionToLoadOpt` is provided, will use the checkpoint pointed to by the _last_checkpoint
+   * file.
+   */
+  private Optional<Long> getStartCheckpointVersion(Engine engine, Optional<Long> versionToLoadOpt) {
```

Comment:
In the future we can also add nice tests for each of these individual getLogSegment util helper methods.

I invision a deep getLogSegment method (as opposed to, basically, getLogSegmentA calls getLogSegmentB which calls getLogSegmentC).

Instead, getLogSegment is 1 large method, and it has little utilities that poke and help out.